jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -524,28 +511,54 @@ void parseAndFillForContent()
 
             assert !_requestBuffer.hasRemaining();
 
+            int filled;
+
+            // The application has retained the content chunks
             if (_requestBuffer.isRetained())
             {
-                // The application has retained the content chunks,
-                // reacquire the buffer to avoid overwriting the content.
-                releaseRequestBuffer();
-                ensureRequestBuffer();
+                // then we must be careful to not overwrite content.
+
+                // If there is space, we can top up the buffer
+                ByteBuffer backing = _requestBuffer.getByteBuffer();
+                if (backing.limit() < backing.capacity() / 8)
+                {
+                    // pad the buffer so retained content is not overwritten
+                    int padding = backing.position();
+                    backing.position(0);
+                    try
+                    {
+                        filled = fillRequestBuffer();
+                    }
+                    finally
+                    {
+                        backing.position(padding);
+                    }
+                }
```

Comment:
@sbordet @lorban This is a new optimization.... I'm guessing you will hate it on first viewing... but please think about it a bit.

If an application is retaining small chunks, this change prevents the wasted memory of having a mostly empty buffer unused.

Currently I have "worked around" this issue by making the RBB Accumulator copy small chunks instead of retaining them, but with this change, it would not be necessary.   This change also works for apps that do not use RBB accumulator.

The change is relevant to this PR, as I'd like to do a delayed dispatch that simply retains chunks

Reply:
I don't hate the idea, really not. It's going to make the buffer logic more error-prone, and there's a good chance that this is going to introduce some buffer leak bug now or in the future, but I really don't mind as 12.0 is resistant to that. And the buffer pool still is a moderate point of contention, so the more we avoid it, the better.

Nevertheless, I'm questioning the heuristic (1/8 sounds awfully like you know the buffer is 8KB) and the implementation (isn't it more likely that the biggest free space is after limit rather than before position?).

It's a bit orthogonal, but I also wonder where is the inflection point where memory copy (of the RBB Accumulator) is cheaper than a CAS (of the RBB retain/release)? I.e.: is there anything to be gained by using the RBB Accumulator more often or not?

Reply:
I would assume that copying less than 16 bytes would probably be cheap... except then we need two buffers rather than one, so it can double the memory commitment.   
So I'm happy to strive for zero copies to keep memory foot print small.